sycl/doc/extensions/proposed/sycl_ext_oneapi_num_compute_units.rst

```diff
@@ -0,0 +1,151 @@
+
+This extension provides a new query, ``info::device::num_compute_units``, with
+the aim to clarify the meaning of "compute units" in SYCL and drive consistency
+across implementations.
+
+
+Specification
+=============
+
+Feature test macro
+------------------
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+``SYCL_EXT_ONEAPI_NUM_COMPUTE_UNITS`` to one of the values defined in the table
+below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's value
+to determine which of the extension's features the implementation supports.
+
+.. table::
+   :align: left
+
+   =====  ===========
+   Value  Description
+   =====  ===========
+   1      Initial version of this extension.
+   =====  ===========
+
+Compute units
+-------------
+
+A SYCL device is divided into one or more compute units, which are each divided
+into one or more processing elements.
+
+All work-items in a given work-group must execute on exactly one compute unit.
+The mapping of work-groups to compute units is not guaranteed: work-groups may
+be dispatched to compute units in any order, and this order may be different
+for every kernel launch.
+
+An implementation may execute multiple work-groups on a single compute unit
+simultaneously, subject to the resource constraints described by other device
+and kernel queries.
+
+The representation of specific hardware architectures in terms of compute units
+is tied to the execution model exposed by an implementation and is thus
+implementation-defined.
+
+[*Note:* To improve the portability of SYCL programs, implementations are
+encouraged to define compute units such that it is possible to fully utilize the
+hardware resources of a device by launching one work-group of size
+``max_work_group_size`` on each compute unit.
+*--end note*]
+
+Device queries
+--------------
+
+This extension adds the following device information descriptor.
+
+----
+
+info::device::num_compute_units
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: c++
+
+   namespace sycl::ext::oneapi::info::device {
+   struct num_compute_units {
+     using return_type = size_t;
+   };
+   }
+
+*Remarks:* Template parameter to ``device::get_info``.
+
+*Returns:* The number of compute units in the device.
+The minimum value is 1.
+
+[*Note:* The value is not required to be equal to the value returned by
+``max_compute_units``.
+*--end note*]
+
+----
+
+
+Implementation in DPC++
+=======================
+
+This section is non-normative and applies only to the DPC++ implementation.
+
+The table below explains how DPC++ calculates the number of compute units for
+different combinations of device and backend.
+
+.. table::
+   :align: left
+
+   ===========  ==========  =================
+   Device Type  Backend(s)  Number of Domains
```

Comment:
@Pennycook: Is "Number of Domains" the right title for this column?  This is what you had in the original Asciidoc version, but it seems like "Number of Compute Units" might be better.

Reply:
Oops.  You're right; I based this extension on the abandoned "execution domains" extension, and missed this.